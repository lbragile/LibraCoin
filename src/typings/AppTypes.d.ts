import { ACTIONS } from "../enums/AppDispatchActions";

export interface ITransaction {
  to: string;
  from: string;
  amount: number | undefined;
  message: string;
  signature: string;
}

export interface IUser {
  publicKey: string;
  balance: number;
}

export interface IBlock {
  index: number;
  prevHash: string;
  currHash: string;
  transactions: ITransaction[];
  timestamp: number;
  merkleRoot: string;
  valid?: boolean;
}
export interface IState {
  verifiedTrans: ITransaction[];
  selectedTrans: ITransaction[];
  users: IUser[];
  chain: IBlock[];
}

export interface IAction {
  type: ACTIONS;
  payload?: unknown;
}
