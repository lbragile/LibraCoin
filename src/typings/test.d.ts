export declare global {
  namespace NodeJS {
    interface Global {
      state: IState;
      dispatch: React.Dispatch<IAction>;
      spki: string;
      spki_hex: string;
      pkcs8: string;
      pkcs8_hex: string;
      privateKey: CryptoKey;
      publicKey: CryptoKey;
      LibraCoinSHA: string;
      LibraCoinEncode: Uint8Array;
      generateKeyMock: jest.Mock;
      exportKeyMock: jest.Mock;
      digestMock: jest.Mock;
      getRandomValuesMock: jest.Mock;
    }
  }
}
