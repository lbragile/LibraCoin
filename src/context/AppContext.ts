import React from "react";
import { IAction, IState } from "../typings/AppTypes";

type TAppContext = { state: IState; dispatch: React.Dispatch<IAction> } | undefined;

export const AppContext = React.createContext<TAppContext>(undefined);
