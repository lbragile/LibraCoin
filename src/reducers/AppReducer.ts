import { ACTIONS } from "../enums/AppDispatchActions";
import { IAction, IState, IUser, ITransaction, IBlock } from "../typings/AppTypes";

export const AppReducer = (state: IState, action: IAction): IState => {
  switch (action.type) {
    case ACTIONS.ADD_VERIFIED_TRANS: {
      const newTrans = (action.payload as { trans: ITransaction }).trans;
      const prevTrans = JSON.parse(localStorage.getItem("verTrans") as string) ?? [];
      const verifiedTrans = [...prevTrans, newTrans];
      localStorage.setItem("verTrans", JSON.stringify(verifiedTrans));
      return { ...state, verifiedTrans };
    }

    case ACTIONS.UPDATE_VERIFIED_TRANS: {
      const selectedTransSignatures = state.selectedTrans.map((x) => x.signature);
      const verifiedTrans = state.verifiedTrans.filter((x) => !selectedTransSignatures.includes(x.signature));
      localStorage.setItem("verTrans", JSON.stringify(verifiedTrans));
      return { ...state, verifiedTrans };
    }

    case ACTIONS.UPDATE_SELECTED_TRANS: {
      const selectedTrans = (action.payload as { selectedTrans: ITransaction[] }).selectedTrans;
      localStorage.setItem("selTrans", JSON.stringify(selectedTrans));
      return { ...state, selectedTrans };
    }

    case ACTIONS.UPDATE_USERS: {
      const users = (action.payload as { users: IUser[] }).users;
      localStorage.setItem("users", JSON.stringify(users));
      return { ...state, users };
    }

    case ACTIONS.ADD_BLOCK: {
      const { block } = action.payload as { block: IBlock };
      const chain = [...state.chain, block];
      localStorage.setItem("chain", JSON.stringify(chain));
      return { ...state, chain };
    }

    case ACTIONS.UPDATE_BLOCK: {
      const { block } = action.payload as { block: IBlock };
      const chain = JSON.parse(localStorage.getItem("chain") as string);
      chain[block.index] = block;
      localStorage.setItem("chain", JSON.stringify(chain));
      return { ...state, chain };
    }

    default:
      return state;
  }
};
