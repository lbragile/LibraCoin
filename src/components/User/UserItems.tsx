import React, { useState, useContext } from "react";
import { Form, InputGroup } from "react-bootstrap";
import { AppContext } from "../../context/AppContext";
import { IAction, IState, IUser } from "../../typings/AppTypes";
import { copyKey } from "../../utils/copyInput";

import "./User.scss";

export default function UserItems(): JSX.Element {
  const { state } = useContext(AppContext) as { state: IState; dispatch: React.Dispatch<IAction> };
  const [copied, setCopied] = useState<boolean[]>(new Array(state.users.length).fill(false));

  function resetCopy(index: number): void {
    const newCopied = JSON.parse(JSON.stringify(copied));
    newCopied[index] = false;
    setCopied(newCopied);
  }

  return (
    <div className="container-fluid mb-2">
      <h3 className="font-weight-bold">Users</h3>
      <div className="row flex-nowrap overflow-auto bg-dark mx-1 px-2 rounded">
        {state.users?.map((user: IUser, i: number) => {
          return (
            <Form className="user-item rounded flex-shrink-0" key={`user-${user.publicKey}`}>
              <InputGroup>
                <InputGroup.Prepend>
                  <InputGroup.Text>ðŸ”‘</InputGroup.Text>
                </InputGroup.Prepend>
                <Form.Control
                  aria-label="User Public Key"
                  type="text"
                  className="text-truncate rounded-right"
                  onFocus={(e: React.FocusEvent<HTMLInputElement>) =>
                    copyKey(e, setCopied, undefined, i, state.users.length)
                  }
                  onBlur={() => resetCopy(i)}
                  defaultValue={user.publicKey}
                  isValid={copied[i]}
                  readOnly
                />
                <Form.Control.Feedback type="valid">Copied to clipboard</Form.Control.Feedback>
              </InputGroup>

              <InputGroup className="mt-2">
                <Form.Control aria-label="balance" type="number" defaultValue={user.balance} disabled />
                <InputGroup.Append>
                  <InputGroup.Text>LC</InputGroup.Text>
                </InputGroup.Append>
              </InputGroup>
            </Form>
          );
        })}
      </div>
    </div>
  );
}
