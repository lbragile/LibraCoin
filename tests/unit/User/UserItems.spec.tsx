/**
 * @group unit
 */

import React, { useReducer } from "react";
import { render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";

import UserItems from "../../../src/components/User/UserItems";
import { AppContext } from "../../../src/context/AppContext";
import { IAction, IState } from "../../../src/typings/AppTypes";
import { AppReducer } from "../../../src/reducers/AppReducer";

const { initialState } = global;

interface IUserItemsWrapper {
  stateMock?: IState;
  dispatchMock?: React.Dispatch<IAction>;
}

const UserItemsWrapper = ({ stateMock, dispatchMock }: IUserItemsWrapper) => {
  const [state, dispatch] = useReducer(AppReducer, initialState);

  return (
    <AppContext.Provider value={{ state: stateMock ?? state, dispatch: dispatchMock ?? dispatch }}>
      <UserItems />
    </AppContext.Provider>
  );
};

it("renders correctly", () => {
  const { asFragment } = render(<UserItemsWrapper />);

  const publicKeys = screen.getAllByRole("textbox", { name: /User Public Key/i });
  const balances = screen.getAllByRole("spinbutton", { name: /balance/i });

  expect(screen.getByRole("heading", { name: /Users/i, level: 3 })).toHaveTextContent("Users");

  balances.forEach((balance, i) => {
    expect(balance).toBeDisabled();
    expect(balance).toHaveValue(initialState.users[i].balance);
  });

  publicKeys.forEach((key, i) => {
    expect(key).toHaveAttribute("readOnly");
    expect(key).toHaveValue(initialState.users[i].publicKey);
  });

  const numUsers = initialState.users.length;
  expect(publicKeys.length).toEqual(numUsers);
  expect(balances.length).toEqual(numUsers);

  expect(asFragment()).toMatchSnapshot();
});

describe("public key of user - copy/blur", () => {
  beforeEach(() => {
    document.execCommand = jest.fn();
    render(<UserItemsWrapper />);
  });

  afterEach(() => {
    jest.restoreAllMocks();
  });

  it("shows 'copied to clipboard' on the correct key when public key field is focused", () => {
    const publicKeyFields = screen.getAllByRole("textbox", { name: /User Public Key/i });
    publicKeyFields.forEach((key) => expect(key).not.toHaveClass("is-valid"));

    publicKeyFields[1].focus();

    expect(publicKeyFields[0]).not.toHaveFocus();
    expect(publicKeyFields[1]).toHaveFocus();
    expect(publicKeyFields[2]).not.toHaveFocus();

    expect(publicKeyFields[0]).not.toHaveClass("is-valid");
    expect(publicKeyFields[1]).toHaveClass("is-valid");
    expect(publicKeyFields[2]).not.toHaveClass("is-valid");
  });

  it("removes 'copied to clipboard' from the correct key when focus is lost", () => {
    const publicKeyFields = screen.getAllByRole("textbox", { name: /User Public Key/i });
    publicKeyFields.forEach((key) => expect(key).not.toHaveClass("is-valid"));

    publicKeyFields[1].focus();

    expect(publicKeyFields[0]).not.toHaveFocus();
    expect(publicKeyFields[1]).toHaveFocus();
    expect(publicKeyFields[2]).not.toHaveFocus();

    expect(publicKeyFields[0]).not.toHaveClass("is-valid");
    expect(publicKeyFields[1]).toHaveClass("is-valid");
    expect(publicKeyFields[2]).not.toHaveClass("is-valid");

    publicKeyFields[1].blur();

    publicKeyFields.forEach((key) => {
      expect(key).not.toHaveFocus();
      expect(key).not.toHaveClass("is-valid");
    });
  });
});
