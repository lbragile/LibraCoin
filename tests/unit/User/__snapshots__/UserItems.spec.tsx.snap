// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<DocumentFragment>
  <div
    class="container-fluid mb-2"
  >
    <h3
      class="font-weight-bold"
    >
      Users
    </h3>
    <div
      class="row flex-nowrap overflow-auto bg-dark mx-1 px-2 rounded"
    >
      <form
        class="user-item rounded flex-shrink-0"
      >
        <div
          class="input-group"
        >
          <div
            class="input-group-prepend"
          >
            <span
              class="input-group-text"
            >
              ðŸ”‘
            </span>
          </div>
          <input
            aria-label="User Public Key"
            class="text-truncate rounded-right form-control"
            readonly=""
            type="text"
            value="3059301306072a8648ce3d020106082a8648ce3d030107034200048fa7d69599babb"
          />
          <div
            class="valid-feedback"
          >
            Copied to clipboard
          </div>
        </div>
        <div
          class="mt-2 input-group"
        >
          <input
            aria-label="balance"
            class="form-control"
            disabled=""
            type="number"
            value="1000"
          />
          <div
            class="input-group-append"
          >
            <span
              class="input-group-text"
            >
              LC
            </span>
          </div>
        </div>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
